### ISSUE TITLE
Up-Arrow behavior different depending on if previous program has finished or not

### ISSUE THREAD

**Issue by greggman** [ohmyzsh/ohmyzsh#6261]
I've noticed that if I press up arrow at a prompt then I get the previous command and up again gets me the command before that.

Whereas if I press up arrow *before the previous program has completed* then instead I get the previous command displayed, the cursor is at the end of the line, but oh-my-zsh is now in "search for lines that start with ... " mode meaning I can't press up to get the previous command.

I'm sure this behavior is well known and expected but just in case you can test like this. 

1. Type `ls` <kbd>return</kbd>  
2. Type `sleep 3` <kbd>return</kbd>
3. wait 3 seconds for prompt to appear
4. press <kbd>up</kbd> (should show `sleep 3`)
5. press <kbd>up</kbd> again (should show `ls`)
6. press <kbd>return</kbd> (to run `ls`)
7. Type `sleep 3` <kbd>return</kbd> <kbd>up</kbd> (press the up arrow before the 3 seconds elapses)
8. It should now be showing `sleep 3`
9. Press <kbd>up</kbd> 
   it will still be showing `sleep 3` but it want it to be showing `ls`. Instead it is in "search for commands that start with `sleep 3` mode instead of just `go to previous command` mode.

To make try to clear in both cases these are the steps

1. `ls`<kbd>return</kbd> 
2. `sleep 3`<kbd>return</kbd> 
3. <kbd>up</kbd> 
4. <kbd>up</kbd> 

But they end up with different results depending on if step3 happens before step2 finished or after.

How do I get oh-my-zsh to be consistent here so that a premature up arrow behaves the same as a normal up arrow? Thank you

**Comment by greggman** [comment]
Does this really not affect anyone but me? I'm really surprised.

**Comment by greggman** [comment]
In any case those looking for a fix you can add this to your .zshrc file sometime after including oh-my-zsh.sh

```
bindkey "^[[A" up-line-or-beginning-search # Up
bindkey "^[[B" down-line-or-beginning-search # Down
```

**Comment by mcornella** [comment]
I can confirm this happens to me too. The reason is that when a command is running, zle is not active and as such the terminal is not in application mode:
https://github.com/robbyrussell/oh-my-zsh/blob/4aa545abedb430df2fa1ae2cc7cc502a5be1a80b/lib/key-bindings.zsh#L5-L16

This means that `$terminfo` sequences don't apply anymore, so the up arrow key runs it's default bindkey widget, which in my case was `up-line-or-history` or `down-line-or-history`. This widget fills the `$BUFFER` (i.e. the command line input) instead of signaling a history search, so when you press the up arrow again, the real widget is run (which in my case was `history-substring-search-up`, in yours, `up-line-or-beginning-search` probably). The widget thinks the command line is written and therefore searches for commands beginning (in your case) or containing (in mine) the command written by `up-line-or-history`.

One possible fix is using the sequence you get when running `cat -v`, which are the ones you wrote in the comment above. #5113 was supposed to fix this but we haven't had a wide enough group of testers to push it forward.

**Comment by robbyrussell** [comment]
We'll consider this resolved should #5113 be merged
